File: Class81_Questions/NumberOfDistinctIslands.java
public class Solution {
    /**
     * @param grid: a list of lists of integers
     * @return: return an integer, denote the number of distinct islands
     */

    public void dfs(int[][] grid, int row, int col, StringBuilder pattern){
        int nRows = grid.length; 
        int nCols = grid[0].length;

        grid[row][col] = 0;

        if(col+1 < nCols && grid[row][col+1] == 1){
            pattern.append("R");
            dfs(grid, row, col+1, pattern);
        }

        if(row+1 < nRows && grid[row+1][col] == 1){
            pattern.append("D");
            dfs(grid, row+1, col, pattern);
        }

        if(col-1 >= 0 && grid[row][col-1] == 1){
            pattern.append("L");
            dfs(grid, row, col-1, pattern);
        }

        if(row-1 >= 0 && grid[row-1][col] == 1){
            pattern.append("U");
            dfs(grid, row-1, col, pattern);
        }

        pattern.append("B");
    }

    public int numberofDistinctIslands(int[][] grid) {
        // write your code here

        Set<String> patternSet = new HashSet<>();      

        for(int i=0; i<grid.length; i++){
            for(int j=0; j<grid[0].length; j++){
                if(grid[i][j] == 1){
                    StringBuilder pattern = new StringBuilder("x");
                    dfs(grid, i, j, pattern);
                    patternSet.add(pattern.toString());
                }
            }
        }

        return patternSet.size();
    }
}

File: Class81_Questions/NumberOfDistinctIslands.cpp
class Solution {
public:
    
    void dfs(vector<vector<int>>& grid, int row, int col, string &pattern){
        grid[row][col] = 2;
        if(col+1 < grid[0].size() && grid[row][col+1] == 1){
            pattern += "r";
            dfs(grid, row, col+1, pattern);
        }
        
        if(row+1 < grid.size() && grid[row+1][col] == 1){
            pattern += "d";
            dfs(grid, row+1, col, pattern);
        }
        
        if(col-1 >= 0 && grid[row][col-1] == 1){
            pattern += "l";
            dfs(grid, row, col-1, pattern);
        }
        
        if(row-1 >= 0 && grid[row-1][col] == 1){
            pattern += "u";
            dfs(grid, row-1, col, pattern);
        }
        
        pattern += "b"; // for backtracking
    }
    
    int numDistinctIslands(vector<vector<int>>& grid) {
        int nRows = grid.size();
        int nCols = grid[0].size();
        
        unordered_set<string> distinctIslands;
        
        for(int i=0; i<nRows; i++){
            for(int j=0; j<nCols; j++){
                if(grid[i][j] == 1){
                    string pattern = "x";
                    dfs(grid, i, j, pattern);
                    distinctIslands.insert(pattern);
                }
            }
        }
        
        return distinctIslands.size();
    }
};